consensus/core/src/transaction_certifier.rs

```diff
@@ -0,0 +1,531 @@
+            context,
+            blocks: BTreeMap::new(),
+            gc_round: GENESIS_ROUND,
+        }
+    }
+
+    fn add_voted_blocks(
+        &mut self,
+        voted_blocks: Vec<(VerifiedBlock, Vec<TransactionIndex>)>,
+    ) -> Vec<CertifiedBlock> {
+        let mut certified_blocks = vec![];
+        for (voted_block, reject_votes) in voted_blocks {
+            let blocks = self.add_voted_block(voted_block, reject_votes);
+            certified_blocks.extend(blocks);
+        }
+        certified_blocks
+    }
+
+    fn add_voted_block(
+        &mut self,
+        voted_block: VerifiedBlock,
+        reject_votes: Vec<TransactionIndex>,
+    ) -> Vec<CertifiedBlock> {
+        let mut certified_blocks = vec![];
+
+        if voted_block.round() <= self.gc_round {
+            // Block is outside of GC bound.
+            return vec![];
+        }
+
+        let block_info = self.blocks.entry(voted_block.reference()).or_default();
+        if block_info.block.is_some() {
+            // Input block has already been processed and added to the state.
+            return vec![];
+        }
+        block_info.block = Some(voted_block.clone());
+        block_info.own_reject_votes = Some(reject_votes.clone());
+
+        // Add own reject votes for the block.
+        for reject in &reject_votes {
+            block_info
+                .reject_votes
+                .entry(*reject)
+                .or_default()
+                .add_unique(self.context.own_index, &self.context.committee);
+        }
+
+        // Add own accept votes for the block.
+        block_info
+            .accept_votes
+            .add_unique(self.context.own_index, &self.context.committee);
+
+        // Check if the input block is now certified.
+        if let Some(certified_block) = block_info.take_certified_output(&self.context.committee) {
+            certified_blocks.push(certified_block);
+        }
+
+        // Update reject votes from the block first. Otherwise another block can get certified when it should not be.
+        for block_votes in voted_block.transaction_votes() {
+            if block_votes.block_ref.round <= self.gc_round {
+                // Block is outside of GC bound.
+                continue;
+            }
+            let block_info = self.blocks.entry(block_votes.block_ref).or_default();
+            for reject in &block_votes.rejects {
+                block_info
+                    .reject_votes
+                    .entry(*reject)
+                    .or_default()
+                    .add_unique(voted_block.author(), &self.context.committee);
+            }
+            if let Some(certified_block) = block_info.take_certified_output(&self.context.committee)
+            {
+                certified_blocks.push(certified_block);
+            }
+        }
+
+        // Update accept votes from the block after updating reject votes.
+        // Only parent round blocks receive accept votes on the fastpath.
```

Comment:
To clarify, on every block proposal we will always have atleast 2f+1 votes to accept all transactions in the block. However it is possible one of those voting blocks includes an explicit reject vote for one transaction. If that happens the only way to get the block certified is either via commit as we can never get 2f+1 accept votes or 2f+1 reject votes right?